examples/graphs/node2vec_movielens.py

```diff
@@ -0,0 +1,584 @@
+"""
+Title: Graph representation learning with node2vec.
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/05/15
+Last modified: 2021/05/15
+Description: Implementing the node2vec to generate embeddings for movies.
+"""
+
+"""
+## Introduction
+
+
+Learning useful representations from objects structured as graphs is useful for
+a variety of ML applications —such as social and communication networks analysis,
+biomedicine studies, and recommendation systems, and leads to greater predictive power.
+[Graph Representation Learning](https://www.cs.mcgill.ca/~wlh/grl_book/) aims to
+learn embeddings for the graph nodes, which can be used for a variety of ML tasks
+such as node label prediction (e.g. categorizing an article based on its citations)
+and link prediction (e.g. recommending an interest group to a user in a social network).
+
+[node2vec](https://arxiv.org/abs/1607.00653) is a simple, yet scalable and effective
+technique for learning low-dimensional embeddings for nodes in a graph by optimizing
+a neighborhood preserving objective. The aim is to learn similar embeddings for
+neighboring nodes, with respect to the graph structure.
+
+Given your data items structured as a graph (where the items are represented as
+nodes and the relationship between items are represented as edges),
+node2vec works as follows:
+
+1. Generate item sequences using (biased) random walk.
+2. Create positive and negative training examples from these sequences.
+3. Train a [word2vec](https://www.tensorflow.org/tutorials/text/word2vec) model
+(skip-gram) to learn embeddings for the items.
+
+In this example, we demonstrate the node2vec technique on the [small version of the Movielens dataset](https://files.grouplens.org/datasets/movielens/ml-latest-small-README.html)
+to learn movie embeddings. Such a dataset can be represented as a graph by treating
+the movies as nodes, and creating edges between movies that have similar ratings
+by the users. The learnt movie embeddings can be used for tasks such as movie recommendation,
+or movie genres prediction.
+
+This example requires `networkx` package, which can be installed using the following command:
+
+```
+pip install networkx
+````
+"""
+
+"""
+## Setup
+"""
+
+import os
+from collections import defaultdict
+import math
+import networkx as nx
+import random
+from tqdm import tqdm
+from zipfile import ZipFile
+from urllib.request import urlretrieve
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+import matplotlib.pyplot as plt
+
+"""
+## Download and Prepare the Data
+
+The small version of the Movielens dataset includes around 100 thousand ratings
+from 610 users on 9742 movies.
+
+First, let's download the movielens data. The downloaded folder will contain
```

Comment:
Suggestion:

```suggestion
First, let's download the dataset. The downloaded folder will contain
```

(to avoid repetition)